ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WSQueryEndpoint.java

```diff
@@ -296,15 +308,15 @@ private void handlePrintTopic(final SessionAndRequest info, final PrintTopic pri
         new WebSocketSubscriber<>(info.session, mapper);
     this.subscriber = topicSubscriber;
 
-    topicPublisher.start(exec, serviceContext.getSchemaRegistryClient(),
+    topicPublisher.start(exec, info.serviceContext.getSchemaRegistryClient(),
```

Comment:
Same comment as the `StreamPublisher`. We need to pass the `serviceContext` as well. The `PrintPublisher` creates a KafkaConsumer without impersonation.

Look at `TopicStreamWriter` where I had to inject the impersonation properties to the KafkaConsumer. Now that we have a `KsqlSecurityExtension`, I think it would be better to pass the security context only and get the consumer and SR from the publisher (which are already impersonated).

Reply:
Done. Thanks for the example. I noticed `PrintPublisher` and `TopicStreamWriter` shared a lot of code so I refactored the two instead of duplicating the code for injecting impersonation properties.